### PR TITLE
Triangulation with simple time differences

### PR DESCRIPTION
**Pull request by tsengj10**
DiffPoint takes simple time differences and points to the supernova. astropy is used to convert the Earth locations of the detectors into geocentric celestial coordinates at the time of observation.

This PR also includes DetectorDB, which reads the detector database file (such as detector_locations.csv) once. Modules can instantiate DetectorDB and then get Detector objects out of it.

GenPoint was a simple SN direction-to-times generator for testing.  Its logic should migrate into the other generator code.

You can test GenPoint/DiffPoint with "make diffpoint".  It will create a mollview skymap.

### CODE REVIEW

File: snewpdag/plugins/DiffPointing.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+          # check if it's already in the cache. If so, delete it.
+          if k in self.cache:
+            self.cache.pop(k)
+          else:
+            krev = (k[1], k[0]) # reverse order
+            if krev in self.cache:
+              self.cache.pop(krev)
+
+    if len(self.cache) >= self.min_dts:
+      return self.reevaluate()
```
Comment (dallaval5u) [inline snewpdag/plugins/DiffPointing.py:192]: I suspect we are missing a data in brackets here

Comment (tsengj10) [reply snewpdag/plugins/DiffPointing.py:192]: I think you're right

**Review by mcolomermolla — COMMENTED** [review]
After a review, and running successfully the code, some comments:
- Config file test-diff.py can be changed to csv for convenience (as for other computations)
- I don't like that time uncertainty is given in the detector_location.csv, it should be given by the experiments in the alert report with T0/time series + expected uncertainty
- I would just improve the map with proper axis labels and also same coordinates convention as for GWs: which corresponds on going from -180 deg to + 180 deg in RA, and you CL should be on the other side (as in Dan plot), mirrowed with respect to the output.

Those changes could be done in a later PR, no objections for merging that one :)

**Comment by tsengj10** [comment]
Thanks @mcolomermolla for the comments.  One question:  didn't Dan's plot have ra=0 at the left edge? Mine has ra=0 at the center (though I have to check that sign again)

**Comment by mcolomermolla** [comment]
If the truth position I run is ra=-60 and dec=-30, and in your plot ra=0 is in the center, the negative RA should be on the left of the plot, not on the right as in the output obtained. The one by Dan is in the right place for ra<0 left ra=0 center and ra>0 right. attached the one I obtained, for the truth position mentioned (close to Galactic Center), it should be "mirrowed":

test-diff-skymap-0-0

**Comment by tsengj10** [comment]
When I look at Dan (et al)'s plot, I see RA of 0h on the left edge, then reading from left to right, 21h, 18h, 15h, all the way back to 0h on the right edge.  I was under the impression that 0h would correspond to 0 degrees, but that might be too naive.

In any case, maybe we should move to use the [ligo.skymap](https://lscsoft.docs.ligo.org/ligo.skymap/index.html) package.

As for the left/right order, this is controllable via the "flip" flag in healpy.mollview.  According to the [healpy.mollview docs](https://healpy.readthedocs.io/en/latest/generated/healpy.visufunc.mollview.html#healpy.visufunc.mollview), the default is the "astro" view, which has "east toward left, west toward right".  The "geo" view is the other way around.  I think this is because the skymap has south pole at the bottom, and the north pole at the top; if you look at the night sky with this orientation, then east is toward the left.  (This seems to be what I get from stellarium, too.)

**Comment by tsengj10** [comment]
Added a csv file for configuration.

It turns out healpy.mollview doesn't offer axis labels; it's apparently a long-standing issue. Probably another reason to move towards ligo.skymap.

Database lookup of sigma and bias removed from DiffPointing. @mcolomermolla is right - it should come from the experiments, or is something we need to calculate ourselves upstream.

**Comment by tsengj10** [comment]
Changed Mollview option to declare "celestial" coordinate system; I had mistakenly put in a "G", which indicates a galactic, sun-centric coordinate system. Doesn't really change the output, because all the calculations were done in the celestial coordinate system.

**Comment by tsengj10** [comment]
As found by @dallaval5u in #61, we were using data['gen'] in two different ways. I've moved truth information to data['truth']

**Comment by tsengj10** [comment]
I think that's it. This should go in because there's more to come in another PR!

**Pull request merged**